src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Private/JavaScriptMarshalImpl.ToCs.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices.JavaScript.Private
+{
+    internal static partial class JavaScriptMarshalImpl
+    {
+        private delegate object MarshalToCsDelegate(IntPtr source);
+        private static Dictionary<IntPtr, MarshalToCsDelegate> toCsMarshallers = new Dictionary<IntPtr, MarshalToCsDelegate>();
+
+#nullable disable
+        internal static unsafe T DynamicMarshalToCs<T>(IntPtr typeHandle, IntPtr source)
```

Comment:
I think we've already discussed some of this, but this method is very slow on paper. There are probably steps we could take to make it efficient, but it might require compile time code generation. We will be relying heavily on inlining and other optimizations as well

Reply:
This dynamic invoke is slow and will be slow because it 
- does lookup in hashtable
- does boxing, to unify the marshaller signature

Generated code would help here. For now I made comments into code about it.